### PR TITLE
return the original string if the list of patterns is empty

### LINKED ISSUE #11485: `String.replace/3` crashes if `pattern` is an empty list
### Environment

* Elixir & Erlang/OTP versions (elixir --version):
```
Erlang/OTP 24 [erts-12.1] [source] [64-bit] [smp:8:8] [ds:8:8:10] [async-threads:1] [jit]
Elixir 1.13.0 (compiled with Erlang/OTP 24)
```
* Operating system: Ubuntu 20.04

### Current behavior
Passing an empty list as the pattern makes the function crash:
```elixir
iex(1)> String.replace("foo", [], "")
** (ArgumentError) errors were found at the given arguments:

  * 2nd argument: not a valid pattern

    (stdlib 3.16) :binary.matches("foo", [])
    (elixir 1.13.0) lib/string.ex:1521: String.replace_guarded/4
```

In my particular use case, the pattern list is generated from an unrelated piece of code, and then passed to `String.replace/3`. Currently I have to do this to make it a no-op on empty lists:
```elixir
updated_text =
  if patterns != [] do
    String.replace(text, patterns, "")
  else
    text
  end
```

### Expected behavior

An empty list of patterns should behave the same way as a pattern that never matches:
```elixir
iex(1)> String.replace("foo", [], "")
"foo"
```

It's not a big deal but I'm not sure if this is by design or not :slightly_smiling_face:

### PR DESCRIPTION
**Pull request by sushant12**
closes #11485

### CODE REVIEW

**Review by eksperimental — CHANGES_REQUESTED** [review]
If we are going this way, it should also be replicated for `String.split/3` which also expects a pattern.

If we Introduce this, I think we should provide a `String.compile_pattern/1` which will deal with versions of pattern that are not handled by `:binary.compile_pattern/1` such as `[]` and `""`.

If we don't introduce this accept these changes, the typespecs for `t:String.pattern/0` needs to be updated to: `t() | nonempty_list(t()) | :binary.cp()`

**Comment by josevalim** [comment]
Hi @sushant12! Thanks for the PR. Unfortunately we decided to not pursue this direction for now, sorry! See the discussion in #11485 for more info.

**Comment by sussbuzz** [comment]
followed every discussion :D . Thanks Jose

File: lib/elixir/test/elixir/string_test.exs [hunk 1]
```diff
@@ -424,6 +424,10 @@ defmodule StringTest do
       assert String.replace("ELIXIR", "", ".", global: false) == ".ELIXIR"
     end
 
+    test "with empty pattern list" do
+      assert String.replace("elixir", [], "") == "elixir"
```
[the remark this answers is not in the document]
Comment (josevalim) [reply lib/elixir/test/elixir/string_test.exs:428]:
```suggestion
      assert String.replace("elixir", [], "anything") == "elixir"
```

**Comment by josevalim** [comment]
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:

**Pull request merged**